engine/crates/engine-v2/src/engine/cache.rs

```diff
@@ -0,0 +1,103 @@
+use base64::{display::Base64Display, engine::general_purpose::URL_SAFE_NO_PAD};
+use engine::PersistedQueryRequestExtension;
+use schema::Schema;
+
+use super::SchemaVersion;
+
+mod namespaces {
+    pub const OPERATION: &str = "op";
+    pub const TRUSTED_DOCUMENT: &str = "tdoc";
+    pub const APQ: &str = "apq";
+}
+
+/// Unique cache key that generates a URL-safe string.
+/// Stable across engine version.
+pub(super) enum Key<'a> {
+    Operation {
+        name: Option<&'a str>,
+        schema_version: &'a SchemaVersion,
+        document: Document<'a>,
+    },
+    TrustedDocument {
+        client_name: &'a str,
+        document_id: &'a str,
+    },
+    Apq {
+        ext: &'a PersistedQueryRequestExtension,
+    },
+}
+
+pub(super) enum Document<'a> {
+    PersistedQueryExt(&'a PersistedQueryRequestExtension),
+    Id(&'a String),
+    Text(&'a str),
+}
+
+impl std::fmt::Display for Key<'_> {
```

Comment:
well did a `Display` because I wanted a `to_string()` and Clippy complained otherwise. You have a fair point, I don't see how I would name it otherwise, to be honest